### ISSUE TITLE
There is no per-benchmark setup when arguments are used

### ISSUE THREAD

**Issue by travisdowns** [google/benchmark#743]
One of the main reasons to set up to a test fixture is to factor out any expensive setup code for your test, so that it runs the minimum number of times.

However, I can't find the right place to put such setup code, assuming the setup depends on the parameters exposed by `State::range()`. Ideally, you'd call the setup once for each parameter combination. If you put it in the fixture constructor however, it will only be called once benchmark registered using that fixture so it's too coarse (and of course as a consequence `State::range()` isn't available there anyways).

If you put it in `Fixture::Setup` it will be called on _every_ benchmark method invocation, including the repeated invocations as the framework searches for the right number of iterations. This means the same data may be repeatedly generated, which also adds noise to the benchmark (e.g., because memory is repeatedly being allocated and de-allocated).

It would be ideal to have a setup method that is called "per-benchmark" (i.e., per line in the tabular output).

**Comment by LebedevRI** [comment]
https://github.com/google/benchmark#basic-usage
> The benchmark library will measure and report the timing for code within the for(...) loop.

So just put it outside (before) of the `for (auto _ : state)` loop.

Does this answer your question?

**Comment by travisdowns** [comment]
@LebedevRI - no, that is essentially the same as putting it in the `Setup` method (plus some extra duplication) - it will be called not once per benchmark, but once every time the benchmark method is invoked (which is 10+ times as the benchmark hunts for right number of iterations).

**Comment by LebedevRI** [comment]
Ah i see.

> If you put it in `Fixture::Setup` it will be called on _every_ benchmark method invocation, including the repeated invocations as the framework searches for the right number of iterations. This means the same data may be repeatedly generated, which also adds noise to the benchmark (e.g., because memory is repeatedly being allocated and de-allocated).

To be noted, the time outside of the loop isn't counted, thus, there will be no direct noise.

But yes, that is not possible. I'm not sure how of a disruptive change it would be to support that.

**Comment by travisdowns** [comment]
> To be noted, the time outside of the loop isn't counted, thus, there will be no direct noise.

Yes, I didn't mean "direct" noise. What I mean is that constantly allocating and deallocating the per-benchmark data adds an unnecessary source of variance. For example, the new allocations may have a different alignment, or no longer be backed by huge pages, happen to alias differently in the L2 cache, end up coming from a different NUMA node, whatever.

Similarly, any setup which uses say a random number generator has to be careful to ensure the generator if freshly seeded with a fixed seed or else you'll be testing different input each time.

**Comment by LebedevRI** [comment]
True.
But then you hit the exact same problem with repetitions.
